### ISSUE TITLE
Tpetra: Thread-parallelize RowMatrixTransposer (for P^T (A*P) in MueLu setup)

### ISSUE THREAD

**Issue by mhoemmen** [trilinos/Trilinos#664]
@trilinos/tpetra @trilinos/muelu 
Story: #629 
Blocked by: #905

Thread-parallelize Tpetra::RowMatrixTransposer.  This matters because of the P^T \* (AP) kernel in MueLu setup.  The transpose should normally still cost less than the sparse matrix-matrix multiply, but once we thread the latter, the transpose may become a bottleneck.  It's also not hard to thread parallelize.

In sparse matrix-matrix multiply, the input Tpetra::CrsMatrix to transpose is always fill complete.  Thus, we can use the KokkosSparse::CrsMatrix directly.  Here's how to get the number of entries in each row of the transpose (result):

```
using Kokkos::Atomic;
using Kokkos::MemoryTraits;
using Kokkos::parallel_for;
using Kokkos::RangePolicy;
Kokkos::View<offset_type*, DT> counts ("counts", lclNumCols);
Kokkos::View<offset_type*, DT, MemoryTraits<Atomic> > countsAtomic = counts;
parallel_for (RangePolicy<typename DT::execution_space, LO> (0, lclNumRows),
  KOKKOS_LAMBDA (const LO& lclRow) {
    for (offset_type k = ptr(lclRow); k < ptr(lclRow+1); ++k) {
      ++countsAtomic(ind(k));
    }
  });
```

**Comment by mndevec** [comment]
Mark, 
I have a kernel written for this purpose in KokkosKernels. But it is using atomic operations, and it runs in 2 phases to count then fill.

```c++
  KOKKOS_INLINE_FUNCTION
  void operator()(const CountTag&, const team_count_member_t & teamMember) const {

    const nnz_lno_t team_row_begin = teamMember.league_rank() * team_work_size;
    const nnz_lno_t team_row_end = KOKKOSKERNELS_MACRO_MIN(team_row_begin + team_work_size, num_rows);

    Kokkos::parallel_for(Kokkos::TeamThreadRange(teamMember,team_row_begin,team_row_end), [&] (const nnz_lno_t& row_index) {
      const size_type col_begin = xadj[row_index];
      const size_type col_end = xadj[row_index + 1];
      const nnz_lno_t left_work = col_end - col_begin;
      Kokkos::parallel_for(
          Kokkos::ThreadVectorRange(teamMember, left_work),
          [&] (nnz_lno_t i) {
        const size_type adjind = i + col_begin;
        const nnz_lno_t colIndex = adj[adjind];
        Kokkos::atomic_fetch_add(&(t_xadj(colIndex)),1);
      });
    });
  }
````

**Comment by srajama1** [comment]
Hmmm.. Do we have to actually transpose this and hold it twice in symmetric case ? Even in unsymmetric case is there a way to avoid the copy ? Say, MueLU building P' is one way.

**Comment by mhoemmen** [comment]
> Hmmm.. Do we have to actually transpose this and hold it twice in symmetric case ? Even in unsymmetric case is there a way to avoid the copy ? Say, MueLU building P' is one way.

Talk to @csiefer2 and @jhux2 about that.

**Comment by jhux2** [comment]
> Hmmm.. Do we have to actually transpose this and hold it twice in symmetric case ? Even in unsymmetric case is there a way to avoid the copy ? Say, MueLU building P' is one way.

@srajama1 In the symmetric case, R=P'.  This means MueLu only needs P and can apply the action of R implicitly.   In the nonsymmetric case, R != P', so MueLu must have both available as separate matrices.

**Comment by srajama1** [comment]
Let me make my question clear: Do you need an explicit transpose at the Tpetra level or KokkosKernels in one of these cases ?

**Comment by jhux2** [comment]
> Let me make my question clear: Do you need an explicit transpose at the Tpetra level or KokkosKernels in one of these cases ?

@srajama1 @csiefer2 Are you asking whether it's sufficient to do a local transpose?

**Comment by srajama1** [comment]
Even more two questions, do you  need a distributed transpose and do you need a local transpose ?

**Comment by jhux2** [comment]
> Even more two questions, do you need a distributed transpose and do you need a local transpose ?

@srajama1 @csiefer2  At the end of the day, we need the action of w=R*v, where w is a multigrid coarse level vector.  At the very least, R must be locally transposed.   If it's not globally transposed, then this would require to do addition work to stitch together the results of applying the locally transposed pieces.

**Comment by csiefer2** [comment]
> Even more two questions, do you need a distributed transpose and do you need a local transpose ?

In general, yes and yes.  But the first is easy.  If we have some code to do local transpose, good ol' transferFillCcomplete can handle the the distributed part.

**Comment by mhoemmen** [comment]
#2236 was a duplicate of this issue.

**Comment by aprokop** [comment]
@mhoemmen Can you summarize what's required to do here? I may need to implement it in the next few days.

**Comment by mhoemmen** [comment]
@aprokop Awesome!  Glad to hear! :-D  Here's what we would need to do

  1. Thread-parallelize `RowMatrixTransposer::createTransposeLocal`
  2. Benchmark to detect any remaining sequential bottlenecks in `expertStaticFillComplete`

The current `createTransposeLocal` implementation works more or less with raw data structures.  The only issue might be whether the input matrix has a valid local matrix yet.  The current code assumes that if the input matrix is a `CrsMatrix`, then `getAllValues` returns valid pointers.  I'm guessing that corresponds to the "fillComplete has been called on it at least once, or it was created with the raw pointers" case.  There is an "input matrix is a RowMatrix not a CrsMatrix" path that we shouldn't need to thread-parallelize.
